data-analytics/dataflow-elasticsearch-indexer/ElasticIndexer/src/main/java/com/google/cloud/pso/utils/PublishProducts.java

```diff
@@ -0,0 +1,262 @@
+         *           are not line delimited. The {@link DoFn} implementation
+         *           generates a separate json message for every product.
+         *        4) Read the product jsons and generate {@link Mutation}
+         *           using {@link JsonToMutationFn}
+         *        5) Write to BigTable
+         *        6) Publish to a PubSub topic.
+         */
+        PCollection<String> jsons =
+                pipeline
+                        //1) Create a filepattern based on the input GCS path.
+                        .apply("Create Filepattern",
+                                Create.of(options.getInput()))
+
+                        // 2) Match the file pattern and read the
+                        //    matched patterns.
+                        .apply(FileIO.matchAll()
+                                .withEmptyMatchTreatment(
+                                        EmptyMatchTreatment.DISALLOW))
+                        .apply(FileIO.readMatches()
+                                .withCompression(Compression.AUTO))
+
+                        // 3) Read files, one per {@link ParDo} as the json
+                        //    messages are not line delimited.
+                        //    The {@link DoFn} implementation generates a
+                        //    separate json message for every product.
+                        .apply("Parse and Extract",
+                                ParDo.of(
+                                        new DoFn<FileIO.ReadableFile, String>() {
+                                            @ProcessElement
+                                            public void processElement(ProcessContext context)
+                                                    throws IOException {
+                                                FileIO.ReadableFile readableFile =
+                                                        context.element();
+                                                String raw = readableFile
+                                                        .readFullyAsUTF8String();
+                                                JSONArray jsonArray =
+                                                        new JSONArray(raw);
+                                                jsonArray.forEach(
+                                                        obj -> {
+                                                            JSONObject jsonObject = (JSONObject) obj;
+                                                            context.output(jsonObject.toString());
+                                                        });
+                                            }
+                                        }));
+
+
+        jsons
+                // 4) Read the product jsons and generate {@link Mutation}
+                //    using {@link JsonToMutationFn}
+                .apply("Create Mutation",
+                        ParDo.of(new JsonToMutationFn(
```

Comment:
This re-parse the JSON which is not performant. Can we have an intermediate data structure so you don't have to parse and re-parse the same records over and over again?

Reply:
I am now passing the JsonNode itself so that the reparsing is avoided when creating a mutation.